python/ray/serve/batching.py

```diff
@@ -0,0 +1,275 @@
+def batch(func: F) -> G:
+    pass
+
+
+# "Decorator factory" use case (called with arguments).
+@overload
+def batch(max_batch_size: int = 10,
+          batch_wait_timeout_s: float = 0.1) -> Callable[[F], G]:
+    pass
+
+
+def batch(_func=None, max_batch_size=10, batch_wait_timeout_s=0.1):
+    """Converts a function to asynchronously handle batches.
+
+    The function can be a standalone function or a class method, and must
+    take a list of objects as its sole argument and return a list of the
+    same length.
+
+    When invoked, the caller passes a single object. These will be batched
+    and executed asynchronously once there is a batch of `max_batch_size`
+    or `batch_wait_timeout_s` has elapsed, whichever occurs first.
+
+    Example:
+        @serve.batch(max_batch_size=50, batch_wait_timeout_s=0.5)
+        async def handle_batch(self, batch: List[str]):
+            return [s.lower() for s in batch]
+
+        async def handle_single(self, s: str):
+            # Will return s.lower().
+            return await handle_batch(s)
+
+    Arguments:
+        max_batch_size (int): the maximum batch size that will be executed in
+            one call to the underlying function.
+        batch_wait_timeout_s (float): the maximum duration to wait for
+            `max_batch_size` elements before running the underlying function.
+    """
+    # `_func` will be None in the case when the decorator is parametrized.
+    # See the comment at the end of this function for a detailed explanation.
+    if _func is not None:
+        if not callable(_func):
+            raise TypeError("@serve.batch can only be used to "
+                            "decorate functions or methods.")
+
+        if not iscoroutinefunction(_func):
+            raise TypeError(
+                "Functions decorated with @serve.batch must be 'async def'")
+
+    if not isinstance(max_batch_size, int):
+        if isinstance(max_batch_size, float) and max_batch_size.is_integer():
+            max_batch_size = int(max_batch_size)
+        else:
+            raise TypeError("max_batch_size must be integer >= 1")
+
+    if max_batch_size < 1:
+        raise ValueError("max_batch_size must be an integer >= 1")
+
+    if not isinstance(batch_wait_timeout_s, (float, int)):
+        raise TypeError("batch_wait_timeout_s must be a float >= 0")
+
+    if batch_wait_timeout_s < 0:
+        raise ValueError("batch_wait_timeout_s must be a float >= 0")
+
+    def _batch_decorator(_func):
+        @wraps(_func)
+        async def batch_wrapper(*args, **kwargs):
+            # Check if this is a method rather than a function by checking
+            # to see if `_func` is the attribute of the first (`self`) argument
+            # under `_func.__name__`. Unfortunately, this is the most robust
+            # solution to this I was able to find. It would also be preferable
+            # to do this check when the decorator runs, rather than when the
+            # method is called.
```

Comment:
can you extract this to a util function, this function is getting quite long

Reply:
I tried hacking around with this and I think it just makes it harder to read / more opaque. Given that we're only doing this in one place I prefer to leave it as-is.

Reply:
oops, I misread where you were talking about originally. pulled this out now.